core/trie/proof.go

```diff
@@ -165,12 +189,329 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
 			if err != nil {
 				return false
 			}
+
+			// Todo:
+			// If we are verifying the key doesn't exist, then we should
+			// update subKey to point in the other direction
+			if value == nil && i == len(proofs)-1 {
+				return true
+			}
+
 			if !proofNode.Edge.Path.Equal(subKey) {
 				return false
 			}
 			expectedHash = proofNode.Edge.Child
 			remainingPath.Truncate(251 - proofNode.Edge.Path.Len()) //nolint:gomnd
 		}
 	}
+
 	return expectedHash.Equal(value)
 }
+
+// VerifyRangeProof verifies the range proof for the given range of keys.
+// This is achieved by constructing a trie from the boundary proofs, and the supplied key-values.
+// If the root of the reconstructed trie matches the supplied root, then the verification passes.
+// If the trie is constructed incorrectly then the root will have an incorrect key(len,path), and value,
+// and therefore it's hash won't match the expected root.
+// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
+func VerifyRangeProof(root *felt.Felt, keys, values []*felt.Felt, proofKeys [2]*Key, proofValues [2]*felt.Felt,
+	proofs [2][]ProofNode, hash hashFunc,
```

Comment:
What is `proofKeys` and `proofValues`? What is the difference from `proofs`?

Reply:
Proofs are the set of nodes from the root to the boundary / proof keys and values

Reply:
I explicitly labelled the proof keys and values for clarity (I find this is a bit easier to understand than placing the proof key as the first element in keys iff the proof key has actually been set - like is done in [geth](https://github.com/ethereum/go-ethereum/blob/master/trie/proof.go#L450))

Reply:
> the left proof should be generated based on the start key query (not actual first leaf that exist)
Currently the left proof key would just correspond to the start key query. And the last proofKey would correspond to the end key query. Similarly with the proof Values (which can be nil if they don't exist)

> Remember, only one proof array, ..
Yup, but that should be a simple matter of merging the proof nodes into a single array, and then unsplitting them again, right? I've opened an [issue](https://github.com/NethermindEth/juno/issues/1898) for this

Reply:
Also, regarding there only being a single proof, geth does store the proof nodes in a key-value database, but I believe there are two boundary proofs, which you can see [here](https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L548-L569). 

Afaiu, when there are key-value pairs to the left and right of the given range we need two boundary proofs. This is because we need to compute the correct shape of the sub-trie, and store the relevant hashes so that we can recompute the root hash for that sub trie. Having only a single boundary proof would no allow us to do this.